chain/indexer-primitives/Cargo.toml

```diff
@@ -0,0 +1,19 @@
+[package]
+name = "near-indexer-primitives"
```

Comment:
Could you add a readme explaining why this crates needs to exist, what should belong to this crate, and what shouldn't belong to this crate? "this crates holds types" doesn't answer those questions for me -- we might as well keep them in indexer, and we also might add some logic here, why not? 

It *seems* to me that what actually happens is that "this crate defines over-the-wire schema for JSON messages used by indexer HTTP API". Ie, it seems that this is not about Rust, but rather about a network protocol/API we expose. That is, if true, a very important bit of information, and it currently is missing from the description. 


:thinking: wait, I think the above is not true. I've just tried removing `Serialize, Deserialize` from all intexer messages, and that seems to work -- nothing in *this* repository seems to rely on serialization. I guess, I am just confused as to why this particular bunch of types needs special treatment and is worth a separate crate :)

Reply:
@matklad I'll add a little README. 

We are making [NEAR Lake](https://github.com/near/near-lake) - in a few words it's an indexer that is going to save all the data from the network to AWS S3 in JSON files. We are starting some NEAR Lake Framework which is going to read from AWS S3 and stream all the data. Both of the mentioned projects will use indexer primitives, in nearest future, there are going to be even more projects. 

The first goal is to make it possible that projects like NEAR Lake Framework and the ones that use it as a dependency won't require an entire `nearcore` in dependencies in order to actually use only those primitives.

And we are going to extend the number of types in indexer primitives and adjust the serde_json parameters to improve the way the types are serialized.

I hope that makes sense. Let me know if you have questions regarding this extraction to a separate crate :)

Reply:
Ah, this makes sense! And yeah, I want to stress that the salient point of this thing then is *not* the Rust types, but rather the JSON schema, and all of the related backwards compatibility concerns.

And yeah, this totally needs to be a separate crate. I wonder though if `-types` or `-views` would be a better public name than `-primitives`. In today's usage `primitives` is essentially a kitchen sink, and it defines much more than just data schema. 

I also wonder if, long term, this should live in some other repository. It feels odd that this defines a data schema, but the actual data isn't used by the repo (so, it's easy for nearcore developers to break this stuff without realizing this). Contrast this with `views`, which also do define the schema, but this schema is than used by the code in this repository (the node's RPC API). It seems nicer if `indexer-framework` defines just the Rust API (without Serialize), and let's the consumers define their own data schema. 

:thinking: yeah, I guess long-term I can see two worlds:

* In one world, "indexer schema" is something defined completely outside of nearcore as a part of the lake framework. In this world, there's a separate `indexer-types` crate, but it lives in the repo which implements the actual indexer. We can also imagine two different lakes by two different providers with different schema
* In another world, we say that "indexer schema"  is something universal, and is as much a part of nearcore as JSON RPC API. In this world, I would expect this to be a *module* in the crate that defines JSON RPC API (so, `near_primitives::views::indexer` today). In this world, while there might be alternative implementations of the lake, we'd expect them to use the same schema.

But yeah, this is long-term, starting with moving this to a separate crate might make sense (but also do consider just creating a views submodule)

Reply:
@matklad TL;DR: There are several things at play, and you properly identified the long-term plans we actually have.

The types defined in this crate are Rust types for Indexer Framework, which is now tightly coupled with nearcore. You don't see any use of these APIs in nearcore repo because it is designed to be a library (you can find the usage in indexer-for-* projects out there), and we want to stabilize this API in the long term in terms of Rust types and while we already have our own top-level structure, there are still quite some types used in depth (e.g. Transaction and ExecutionOutcome types). Given there is an initiative to decouple JSON RPC types #5516, we decided to postpone solving all the problems.

Second, indeed, we want to be able to serialize the types for near-lake and deserialize it on the reader side without pulling the whole nearcore (we could have get away with pulling the whole nearcore, but due to #3803 it makes things a bit harder on our side, since we often receive questions about M1 support in our tooling and we use M1 ourselves)

Third, there is a plan to further solidify the APIs and extract near-indexer into a separate repo, but to make that happen, nearcore should be published as a crate, and its APIs also stable.